rosbag2_py/src/rosbag2_py/_transport.cpp

```diff
@@ -218,24 +227,38 @@ class Recorder
     }
 
     auto writer = rosbag2_transport::ReaderWriterFactory::make_writer(record_options);
-    auto recorder = std::make_shared<rosbag2_transport::Recorder>(
+    recorder_ = std::make_shared<rosbag2_transport::Recorder>(
       std::move(writer), storage_options, record_options, node_name);
-    recorder->record();
+    recorder_->record();
 
-    exec_->add_node(recorder);
-    // Release the GIL for long-running record, so that calling Python code can use other threads
-    {
+    exec_->add_node(recorder_);
+    try {
+      // Release the GIL for long-running record, so that calling Python code can use other threads
       py::gil_scoped_release release;
-      exec_->spin();
+      while (!exit_) {
+        exec_->spin_all(std::chrono::milliseconds(10));
+      }
+      if (recorder_) {
+        recorder_->stop();
+        // Need to spin node to be able to send notifications about bag split and close
+        exec_->spin_all(std::chrono::milliseconds(10));
+      }
+    } catch (...) {
+      exec_->remove_node(recorder_);
     }
+    exec_->remove_node(recorder_);
   }
 
-  void cancel()
+  static void cancel()
```

Comment:
This still means that we can't have multiple `Recorder` in a single process, correct? I'm not sure if that was really supported before, but this decision pushes away from that possibility.

Reply:
Well, multiple recorders from Python are unlikely since we have `rclcpp::init()` in the constructor and `rclcpp::shutdown()` in the destructor. And it was not able before. 
I only changed the `cancel()` method to be static to be able to call it from signal handlers and it does make sense to me.

If the user presses `CTRL+C` we will stop all recorders in the current thread to be able to gracefully finish and exit from the process.